integration/spark/src/main/scala/org/apache/spark/util/PartitionCacheManger.scala

```diff
@@ -0,0 +1,143 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.net.URI
+import java.util
+
+import scala.collection.JavaConverters._
+
+import org.apache.log4j.Logger
+import org.apache.spark.sql.catalyst.catalog.{CatalogStorageFormat, CatalogTablePartition}
+
+import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.cache.{Cache, Cacheable, CarbonLRUCache}
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.datastore.impl.FileFactory
+import org.apache.carbondata.core.metadata.SegmentFileStore
+import org.apache.carbondata.core.statusmanager.SegmentStatusManager
+import org.apache.carbondata.core.util.path.CarbonTablePath
+
+object  PartitionCacheManager extends Cache[PartitionCacheKey, CacheablePartitionSpec] {
+
+  private val CACHE = new CarbonLRUCache(
+    CarbonCommonConstants.CARBON_PARTITION_MAX_DRIVER_LRU_CACHE_SIZE,
+    CarbonCommonConstants.CARBON_MAX_LRU_CACHE_SIZE_DEFAULT)
+
+  val LOGGER: Logger = LogServiceFactory.getLogService(this.getClass.getName)
+
+  def get(identifier: PartitionCacheKey): CacheablePartitionSpec = {
+    val cacheablePartitionSpec =
+      CACHE.get(identifier.tableId).asInstanceOf[CacheablePartitionSpec]
+    val tableStatusModifiedTime = FileFactory
+      .getCarbonFile(CarbonTablePath.getTableStatusFilePath(identifier.tablePath))
+      .getLastModifiedTime
+    if (cacheablePartitionSpec != null) {
+      if (tableStatusModifiedTime > cacheablePartitionSpec.timestamp) {
+        readPartitions(identifier, tableStatusModifiedTime)
+      } else {
+        cacheablePartitionSpec
+      }
+    } else {
+      readPartitions(identifier, tableStatusModifiedTime)
+    }
+  }
+
+  override def getAll(keys: util.List[PartitionCacheKey]):
+  util.List[CacheablePartitionSpec] = {
+    keys.asScala.map(get).toList.asJava
+  }
+
+  override def getIfPresent(key: PartitionCacheKey): CacheablePartitionSpec = {
+    CACHE.get(key.tableId).asInstanceOf[CacheablePartitionSpec]
+  }
+
+  override def invalidate(partitionCacheKey: PartitionCacheKey): Unit = {
+    CACHE.remove(partitionCacheKey.tableId)
+  }
+
+  private def readPartitions(identifier: PartitionCacheKey, tableStatusModifiedTime: Long) = {
```

Comment:
now it will reload all .segment files again. can we support incremental loading?

Reply:
Do you mean caching of partitions after load?

Reply:
"readPartitions" will read all .segment files after loading. better to load the new .segment file only.

Reply:
ok got it

Reply:
please check query flow. it also use incremental loading index.
here maybe can reuse the segment info cache.

Reply:
added per segment modification check...now only the updated/new segments would be loaded

Reply:
@QiangCai each load or query would be loading the already success segments, so it will now solve the problem you mentioned